### ISSUE TITLE
Behavior of setHandler is not consistent

### ISSUE THREAD

**Issue by ctron** [eclipse-vertx/vert.x#3317]
### Version

3.8.1

### Do you have a reproducer?

```java
        Promise<String> p = Promise.promise();

        p.complete("Foo");

        p.future()
                .setHandler(ar -> {
                    System.out.format("1: %s%n", ar.result());
                })
                .setHandler(ar -> {
                    System.out.format("2: %s%n", ar.result());
                });
```

vs:

```java
        Promise<String> p = Promise.promise();

        p.future()
                .setHandler(ar -> {
                    System.out.format("1: %s%n", ar.result());
                })
                .setHandler(ar -> {
                    System.out.format("2: %s%n", ar.result());
                });

        p.complete("Foo");
```

### Steps to reproduce

1. ... Run the two samples above

### Extra

When the "complete" method is called after setting up the two handlers, only one handler is being called:

```
2: Foo
```

When complete is called before setting up the handler, both handlers get called:

```
1: Foo
2: Foo
```

My expectation would be that, now matter when the future gets completed, always the same handlers get called.

**Comment by vietj** [comment]
@ctron can you elaborate why you are calling several times setHandler on the Future ?

**Comment by vietj** [comment]
a proposal would be that we have in 3.9.0 setHandler to set handler and overwrite any `onXXX` method and we deprecate it as well in 3.9.0 in favor of `onComplete` for removal in 4.0

**Comment by ctron** [comment]
I am using `setHandler` because it does allow you to get called in any case, not matter what the outcome is. Like Java's `whenComplete` or `handle`.

Also, passing along Futures, your code has no way to understand if the someone else already has called `setHandler`, which basically makes `setHandler` a thing you can't trust.

**Comment by vietj** [comment]
I think that Java's whenComplete will get your handler call twice as well.

The main difference between before and now is that `setHandler` is idempotent and `onComplete` is not and that is a regression.

We get the idempotent back in 3.9.0 and we will remove the `setHandler` method in 4.0 + deprecation it in 3.9.0.

**Comment by ctron** [comment]
> I think that Java's whenComplete will get your handler call twice as well.

It does, yet.

> The main difference between before and now is that `setHandler` is idempotent and `onComplete` is not and that is a regression.
> 
> We get the idempotent back in 3.9.0 and we will remove the `setHandler` method in 4.0 + deprecation it in 3.9.0.

I am sorry, but I can't follow you. I didn't use any `onComplete` method. As of vertx 3.8, it doesn't seem to exist.

My expectation would be that, when using `setHandler` I either get two invocations, or only one. But no matter **when** the future completes. I always get the same behavior.

**Comment by vietj** [comment]
sorry for misunderstanding you, so if you use latest version from 3.8 branch, you will get this exact behaviour.

**Comment by ctron** [comment]
So you are saying this issue got fixed for 3.8.x? If that is the case, could you provide the information in which version this is fixed?

**Comment by vietj** [comment]
This commit introduced the feature, it is a backport from Vert.x 4

https://github.com/eclipse-vertx/vert.x/commit/8feaa77803a67d381427d2192f84ac1f2f695cd6 <https://github.com/eclipse-vertx/vert.x/commit/8feaa77803a67d381427d2192f84ac1f2f695cd6>

it is available since 3.8.4
